pkg/apis/servicecatalog/plan_reference.go

```diff
@@ -0,0 +1,167 @@
+	}
+
+	return ""
+}
+
+// GetSpecifiedPlan returns the user-specified plan value from one of:
+// * ClusterServicePlanExternalName
+// * ClusterServicePlanExternalID
+// * ClusterServicePlanName
+// This method is intended for presentation purposes only.
+func (pr PlanReference) GetSpecifiedPlan() string {
+	if pr.ClusterServicePlanExternalName != "" {
+		return pr.ClusterServicePlanExternalName
+	}
+
+	if pr.ClusterServicePlanExternalID != "" {
+		return pr.ClusterServicePlanExternalID
+	}
+
+	if pr.ClusterServicePlanName != "" {
+		return pr.ClusterServicePlanName
+	}
+
+	return ""
+}
+
+// GetClassFilterFieldName returns the appropriate field name for filtering
+// a list of service catalog classes by the PlanReference.
+func (pr PlanReference) GetClassFilterFieldName() string {
+	if pr.ClusterServiceClassExternalName != "" {
+		return "spec.externalName"
+	}
+
+	if pr.ClusterServiceClassExternalID != "" {
+		return "spec.externalID"
+	}
+
+	return ""
+}
+
+// GetPlanFilterFieldName returns the appropriate field name for filtering
+// a list of service catalog plans by the PlanReference.
+func (pr PlanReference) GetPlanFilterFieldName() string {
+	if pr.ClusterServicePlanExternalName != "" {
+		return "spec.externalName"
+	}
+
+	if pr.ClusterServicePlanExternalID != "" {
+		return "spec.externalID"
+	}
+
+	return ""
+}
+
+// String representation of a PlanReference
+// Example: class_name/plan_name, class_id/plan_id
+func (pr PlanReference) String() string {
+	return fmt.Sprintf("%s/%s", pr.GetSpecifiedClass(), pr.GetSpecifiedPlan())
+}
+
+// Format the PlanReference
+// %c - Print specified class fields only
+//    Examples:
+//     {ClassExternalName:"foo"}
+//     {ClassExternalID:"foo123"}
+//     {ClassName:"k8s-foo123"}
+// %b - Print specified plan fields only
+//    NOTE: %p is a reserved verb so we can't use it, and go vet fails for non-standard verbs
+//    Examples:
+//     {PlanExternalName:"bar"}
+//     {PlanExternalID:"bar456"}
+//     {PlanName:"k8s-bar456"}
+// %s - Print a short form of the plan and class
+//    Examples:
+//     foo/bar
+//     foo123/bar456
+//     k8s-foo123/k8s-bar456
+// %v - Print all specified fields
+//    Examples:
+//     {ClassExternalName:"foo", PlanExternalName:"bar"}
+//     {ClassExternalID:"foo123", PlanExternalID:"bar456"}
+//     {ClassName:"k8s-foo123", PlanName:"k8s-bar456"}
+func (pr PlanReference) Format(s fmt.State, verb rune) {
+	classFields := make([]string, 0, 3)
+	if pr.ClusterServiceClassExternalName != "" {
+		classFields = append(classFields, fmt.Sprintf("ClassExternalName:%q", pr.ClusterServiceClassExternalName))
+	}
+	if pr.ClusterServiceClassExternalID != "" {
+		classFields = append(classFields, fmt.Sprintf("ClassExternalID:%q", pr.ClusterServiceClassExternalID))
+	}
+	if pr.ClusterServiceClassName != "" {
+		classFields = append(classFields, fmt.Sprintf("ClassName:%q", pr.ClusterServiceClassName))
+	}
+
+	planFields := make([]string, 0, 3)
```

Comment:
Seems like it'd make sense to just create a `[]string{}` and append to it as necessary. preallocating memory seems like an optimization we don't need (and we might not use any of it at all). am I missing something about how this works?

Reply:
You're right, I just did that so y'all would think I was cool. 😀  Since only 1 at most is usually ever set, I'll switch to an empty slice.